python-sdk/nuscenes/predict/models/mtp.py

```diff
@@ -0,0 +1,317 @@
+                                resnet101, resnet152)
+
+from nuscenes.eval.predict.config import PredictionConfig
+
+
+def trim_network_at_index(network: nn.Module, index: int) -> nn.Module:
+    """
+    Returns a new network with all layers up to index from the back.
+    :param network: Module to trim
+    :param index: Where to trim the network. Counted from the last layer.
+    """
+    return nn.Sequential(*list(network.children())[:-1])
+
+RESNET_VERSION_TO_MODEL = {'resnet18': resnet18, 'resnet34': resnet34,
+                           'resnet50': resnet50, 'resnet101': resnet101,
+                           'resnet152': resnet152}
+
+class ResNetBackbone(nn.Module):
+    """
+    Outputs tensor after last convolution before the fully connected layer.
+
+    Allowed versions: resnet18, resnet34, resnet50, resnet101, resnet152.
+    """
+
+    def __init__(self, version: str):
+        """
+        Inits ResNetBackbone
+        :param vesion: resnet version to use.
+        """
+        super().__init__()
+
+        if version not in RESNET_VERSION_TO_MODEL:
+            raise ValueError(f'Parameter version must be one of {list(RESNET_VERSION_TO_MODEL.keys())}'
+                             f'. Received {version}.')
+
+        self.backbone = trim_network_at_index(RESNET_VERSION_TO_MODEL[version](), -1)
+
+    def forward(self, input_tensor):
+        """
+        Outputs features after last convolution.
+        :param input_tensor:  Shape [batch_size, n_channels, length, width]
+        :return: Tensor of shape [batch_size, n_convolution_filters]. For resnet50,
+            the shape is [batch_size, 2048]
+        """
+        backbone_features = self.backbone(input_tensor)
+        return torch.flatten(backbone_features, start_dim=1)
+
+class MobileNetBackbone(nn.Module):
+    """
+    Outputs tensor after last convolution before the fully connected layer.
+
+    Allowed versions: mobilenet_v2.
+    """
+
+    def __init__(self, version: str):
+        """
+        Inits MobileNetBackbone
+        :param version: mobilenet version to use.
+        """
+        super().__init__()
+
+        if version != 'mobilenet_v2':
+            raise NotImplementedError(f'Only mobilenet_v2 has been implemented. Received {version}.')
+
+        self.backbone = trim_network_at_index(mobilenet_v2(), -1)
+
+    def forward(self, input_tensor):
+        """
+        Outputs features after last convolution.
+        :param input_tensor:  Shape [batch_size, n_channels, length, width]
+        :return: Tensor of shape [batch_size, n_convolution_filters]. For mobilenet_v2,
+            the shape is [batch_size, 1280]
+        """
+        backbone_features = self.backbone(input_tensor)
+        return backbone_features.mean([2, 3])
+
+
+class MTP(nn.Module):
+    """Implements the MTP network."""
+
+    def __init__(self, backbone: nn.Module, num_modes: int,
+                 seconds: int = 6, frequency_in_hz: int = 2,
```

Comment:
Should `seconds` and `frequencey_in_hz` be `int`s?

Reply:
Not necessarily but `seconds * frequency_in_hz` should be an int